src/js/views/Projects/Projects.jsx

```diff
@@ -64,50 +164,73 @@ function Projects() {
   }, [])
 
   useEffect(() => {
-    const url = buildURL()
-    if (
-      state.fetching === false &&
-      (state.lastRequest === null || state.lastRequest.toString() !== url)
-    ) {
+    if (state.fetching === false) {
       setState({ ...state, fetching: true })
-      httpGet(
-        globalState.fetch,
-        url,
-        (result) => {
-          const stateURL = buildURL('/ui/projects')
-          dispatch({
-            type: 'SET_CURRENT_PAGE',
-            payload: {
-              url: stateURL,
-              title: 'projects.title'
-            }
-          })
-          navigate(`${stateURL.pathname}?${stateURL.searchParams.toString()}`)
-          setState({
-            ...state,
-            data: result.data,
-            fetching: false,
-            lastRequest: url,
-            rowCount: result.rows
-          })
-        },
-        (error) => {
-          setErrorMessage(t('projects.requestError', { error: error }))
-          setState({
-            ...state,
-            fetching: false,
-            filter: {
-              include_archived: false,
-              namespace_id: null,
-              project_type_id: null
-            },
-            lastRequest: url,
-            sort: { name: 'asc' }
-          })
+
+      let filter = globalState.projects.filter
+      if (
+        globalState.projects.filter === '' ||
+        globalState.projects.filter === '*'
+      )
+        filter = 'NOT archived:true'
+      else if (
+        globalState.projects.filter.length > 0 &&
+        globalState.projects.filter.includes('archived:') !== true
+      )
+        filter = `(${globalState.projects.filter}) AND NOT archived:true`
+
+      const query = {
+        query: toElasticsearchQuery(
+          fromKueryExpression(filter),
+          metadataAsOptions.openSearch
+        ),
+        fields: globalState.projects.fields,
+        size: 1000
+      }
+
+      httpPost(globalState.fetch, buildURL('/opensearch/projects'), query).then(
+        ({ data, success }) => {
+          if (success === true) {
+            const tableData = []
+            data.hits.hits.forEach((row) => {
+              const values = {}
+              for (const [key, value] of Object.entries(row.fields)) {
+                if (slugToNameMap[key] !== undefined) {
+                  values[key] = slugToName(slugToNameMap[key], value[0])
+                } else {
+                  values[key] = value[0]
+                }
+              }
+              tableData.push(values)
+            })
+
+            setState({
+              ...state,
+              data: sortTableData(tableData, globalState.projects.sort),
+              fetching: false,
+              refresh: false
+            })
+          } else {
+            setErrorMessage(t('projects.requestError', { error: data }))
+            setState({
+              ...state,
+              fetching: false,
+              refresh: false
```

Comment:
I'd recommend using the functional `setState` here, since this is called from an async callback. If the state has been updated in the meantime, it's going to get reverted unintentionally.

Reply:
I don't think there really is enough complexity in what's setting state and when you can change it for this to practically be a real issue, but I guess it's a good practice to follow.

Reply:
You're probably right that it's not complex enough, but I'd rather just not have to worry about it at all which streamlines the development and review process because you don't need to then cross-check all other state changes for potential race conditions.

Regarding your change, I don't think that solves the problem. The functions you defined still close over the original `state` variable. See [functional updates](https://reactjs.org/docs/hooks-reference.html#functional-updates) where the arg to `setState` is a function that must return the new state:

```javascript
setState((prevState) => ({
  ...prevState,
  fetching: false,
  refresh: true
}))
```

When React processes this state update `prevState` will be the most up-to-date state.